Add readme (#14)
KULeuven-MICAS/snax-streamer: README.md

# Flexible Streamer Generator for SNAX

> The Flexible Streamer Generator is built to alleviate the data layout concern of an accelerator. It decouples the the accelerator with the real memory system. It simplifies the data interface of the accelerator. The accelerator doesn't need to consider the complicated request and response to/from the real memory system. The accelerator's data interface is simply a valid-ready decoupled handshake interface. Besides, the pre-fetch mechanism inside the Streamer also helps to relieve the data contention problem.

Reviewer: ```suggestion
The Flexible Streamer Generator is a tool to generate _Streamers_ for an accelerator. A Streamer is a reusable hardware block designed to handle the memory accesses, decoupling the accelerator from the actual memory system, and simplifying the interface to the accelerator. The accelerator's data interface is a simple valid-ready handshake interface. On to of this, the pre-fetch mechanism inside the Streamer also helps to relieve data contention in the memory.
```

Reviewer 2: What does 

> On to of this, the pre-fetch mechanism inside the Streamer also helps to relieve data contention in the memory.

Mean @jorendumoulin

> The Flexible Streamer Generator is built to alleviate the data layout concern of an accelerator. It decouples the the accelerator with the real memory system. It simplifies the data interface of the accelerator. The accelerator doesn't need to consider the complicated request and response to/from the real memory system. The accelerator's data interface is simply a valid-ready decoupled handshake interface. Besides, the pre-fetch mechanism inside the Streamer also helps to relieve the data contention problem.

Author: > Why not name it as a SNAX streamer? haha for branding purposes?

* It is used for SNAX but it can be used in general cases too. I found ETH also has its `hwpe-stream` and UCB has its `memstreamer`. But actually, our streamer can cover their functionality to some extent. I am thinking about extending this streamer to a more general one later... (in the long term, maybe).
* I think `Flexible Streamer Generator for SNAX`  is a little bit long!

> The Flexible Streamer Generator is built to alleviate the data layout concern of an accelerator. It decouples the the accelerator with the real memory system. It simplifies the data interface of the accelerator. The accelerator doesn't need to consider the complicated request and response to/from the real memory system. The accelerator's data interface is simply a valid-ready decoupled handshake interface. Besides, the pre-fetch mechanism inside the Streamer also helps to relieve the data contention problem.

Reviewer 2: This section and all of these explanations are rather complicated, can we make it a bit "dumber" in the beginning and delay the complex stuff a bit past the intro?
Something like:

The streamer converts a programmable access pattern to a sequence of addresses in space and time to stream data to and from a banked memory system.

And maybe then lead with your SIMD example @xiaoling-yi ?

> The Streamer has a compatible interface with the [SNAX core](https://github.com/KULeuven-micas/snitch_cluster) and can be integrated into it as a data streaming engine. This repository contains the chisel sources and unit tests for the Flexible Streamer Generator.

Reviewer: ```suggestion
The Streamer has a compatible interface with the [SNAX core](https://github.com/KULeuven-micas/snitch_cluster) and can be integrated into it as a data streaming engine.
```

Thanks to the strong expression power of Chisel, the Flexible Streamer Generator has the capability of dealing with any number of temporal loops and any number of parallel loops for data address generation. More importantly, the Streamre generator is accelerator agnostic, highly flexible, and parameterizable which means it is suitable for a range of accelerators with [tiled-strided-layout](https://github.com/KULeuven-MICAS/snax-mlir/tree/main/compiler/ir/tsl). We have generated Streamers for three accelerators, including the SNAX-GEMM, SNAX-PostProcessing-SIMD, and MAC Engine.

## Microarchitecture
The microarchitecture of the Flexible Streamer Generator is shown below.

The main module of the Streamer is the data mover, including the data reader (reads data from the real memory system and acts as the producer for the accelerator) and data writer (writes data to the real memory system and acts as the consumer for the accelerator). Each data mover has its own address generation unit and works independently so that each data mover can do the data transfer as quickly as possible.

> The data reader gives read requests to the memory system (the address, and read signal, etc.) and gets the response. When all the responses of one transaction are obtained, it pushes the valid data into a FIFO for the accelerator to consume. If the accelerator is consumed successfully, the FIFO will pop the data. The data reader will keep getting new data from the real memory system (pre-fetch) until the FIFO is full or all the input data has been fetched.

Reviewer: ```suggestion
The data reader gives read requests to the memory system (the address, and read signal, etc.) and gets the response. When all the responses of one transaction are obtained, it pushes the valid data into a FIFO for the accelerator to consume. If the accelerator is consumes the data successfully, the FIFO will pop the data. The data reader will keep getting new data from the real memory system (pre-fetch) until the FIFO is full or all the input data has been fetched.
```

|  | dataReaderParams | A sequence of the parameters for dataReader. Each dataReader parameter can be different. The number of the parameters can also be any, meaning there is no constraint on how many the dataReader should be. |
|  | dataWriterParams | Similar as dataReaderParams. A sequence of the parameters for dataWriter. Each dataWriter parameter can be different. The number of the parameters can also be any, meaning there is no constraint on how many the dataWriter should be. |
| | fifoReaderParams | A sequence of the parameters for the FIFOs for dataReaders. The sequence length should matches the length of dataReaderParams.|
| | fifoWriterParams | A sequence of the parameters for the FIFOs for dataWriters. The sequence length should matches the length of dataWriterParams.|

*TCDM is the memory that the Streamer interacts to.

### Instantiate Streamer for three accelerators
`StreamerInstanceParameters.scala` lists three Streamer parameter configurations for three different accelerators.

#### Example: Streamer parameter configuration for a simple MAC Engine

> ```

Reviewer 2: This adds syntax highlighting in this block
```suggestion
```scala
```

> #### Example: Streamer parameter configuration for a simple MAC Engine

Author: Oh! Sound difficult

    FIFOParams(64 * MacScalingFactor, 2),
    FIFOParams(64 * MacScalingFactor, 2),
    FIFOParams(64, 2)
  )

  def fifoWriterParams: Seq[FIFOParams] = Seq(
    FIFOParams(64 * MacScalingFactor, 2)
  )

  def stationarity = Seq(0, 0, 1, 1)

  def dataReaderParams: Seq[DataMoverParams] = Seq(
    DataMoverParams(
      tcdmPortsNum = 1 * MacScalingFactor,
      spatialBounds = Seq(2 * MacScalingFactor),
      spatialDim = 1,
      elementWidth = 32,
      fifoWidth = fifoReaderParams(0).width
    ),
    DataMoverParams(
      tcdmPortsNum = 1 * MacScalingFactor,
      spatialBounds = Seq(2 * MacScalingFactor),
      spatialDim = 1,
      elementWidth = 32,
      fifoWidth = fifoReaderParams(1).width
    ),
    DataMoverParams(
      tcdmPortsNum = 1,
      spatialBounds = Seq(2),
      spatialDim = 1,
      elementWidth = 32,
      fifoWidth = fifoReaderParams(2).width
    )
  )

  def dataWriterParams: Seq[DataMoverParams] = Seq(
    DataMoverParams(
      tcdmPortsNum = 1 * MacScalingFactor,
      spatialBounds = Seq(2 * MacScalingFactor),
      spatialDim = 1,
      elementWidth = 32,
      fifoWidth = fifoWriterParams(0).width
    )
  )

}
```
## Functional description

> Based on the definition of [tiled-strided-layout](https://github.com/KULeuven-MICAS/snax-mlir/tree/main/compiler/ir/tsl), the address generation is based on two parts, the temporal address and spatial address. The temporal address is based on the temporal loop counters and the temporal strides configuration. The spatial address is based on the spatial unrolling factor of the accelerator and the spatial strides configuration. As the accelerator has spatial unrolling, the address generation unit will generate the address for each data element. But when sending the request to the real memory system, these addresses will be merged for data alignment for the memory bank width.

Reviewer 3: ```suggestion
Based on the definition of [tiled-strided-layout](https://github.com/KULeuven-MICAS/snax-mlir/tree/main/compiler/ir/tsl), the address generation has two parts, the temporal address and spatial address. The temporal address is dependent on the temporal loop counters and the temporal strides configuration. The spatial address is dependent on the spatial unrolling factor of the accelerator and the spatial strides configuration. As the accelerator has spatial unrolling, the address generation unit will generate the address for each data element. But when sending the request to the real memory system, these addresses will merge for data alignment of the memory bank width.
```

> Based on the definition of [tiled-strided-layout](https://github.com/KULeuven-MICAS/snax-mlir/tree/main/compiler/ir/tsl), the address generation is based on two parts, the temporal address and spatial address. The temporal address is based on the temporal loop counters and the temporal strides configuration. The spatial address is based on the spatial unrolling factor of the accelerator and the spatial strides configuration. As the accelerator has spatial unrolling, the address generation unit will generate the address for each data element. But when sending the request to the real memory system, these addresses will be merged for data alignment for the memory bank width.

Reviewer: ```suggestion
The access pattern for the Streamer is represented by a set of nested for loops. Every for loop can specify a _stride_ which we use to increment a base pointer. On top of this, some of the for loops can be _spatially unrolled_, to enable parallel data accesses by the accelerator. This splits address generation in two parts, the temporal address generation and spatial address generation. The temporal address is based on the temporal loop counters and the temporal strides configuration. The spatial address is based on the spatial unrolling factor of the accelerator and the spatial strides configuration. The spatial address generation unit will generate a unique address for each data element which is accessed in parallel. As there may be multiple data elements per memory word, these addresses are then merged for proper data alignment with the memory bank width.
```

We formulate the address generation formula below:
```
for(;;temporalbound_n-1)
 for(;;temporalbound_n-2)
   …
    for(;;temporalbound_0)
     parfor(;;spatialbound_m-1)
      parfor(;;spatialbound_m-2)
       …
        parfor(;;spatialbound_0)

>           add_o = base_ptr + temporal_address + spatial_address
>           temporal_address = (temporal_unrolling_loop_counters.zip(tempStride).map { case (a, b) => a * b }).reduce(_ +& _)
>           spatial_address = (spatial_unrolling_loop_counters.zip(spatialStride).map { case (a, b) => a * b }).reduce(_ +& _)

Reviewer: ```suggestion
          temporal_address = (temporal_unrolling_loop_counters.zip(tempStride).map { case (a, b) => a * b }).reduce(_ +& _)
          spatial_address = (spatial_unrolling_loop_counters.zip(spatialStride).map { case (a, b) => a * b }).reduce(_ +& _)
          add_o = base_ptr + temporal_address + spatial_address
```

      addr_o[Vu-1:0] = base_ptr + ti*(tempStride) +  [Vu-1:0]*(spatialStride)
```
Which actually translates to:
```
addr_o[0] = base_ptr + ti*(tempStride) +  [0]*(spatialStride)
addr_o[1] = base_ptr + ti*(tempStride) +  [1]*(spatialStride)
addr_o[2] = base_ptr + ti*(tempStride) +  [2]*(spatialStride)
addr_o[3] = base_ptr + ti*(tempStride) +  [3]*(spatialStride)
addr_o[4] = base_ptr + ti*(tempStride) +  [4]*(spatialStride)
addr_o[5] = base_ptr + ti*(tempStride) +  [5]*(spatialStride)
addr_o[6] = base_ptr + ti*(tempStride) +  [6]*(spatialStride)
addr_o[7] = base_ptr + ti*(tempStride) +  [7]*(spatialStride)
```
Take GEMM Accelerator input data A as an example(3 temporal loop and 2 spatial loop), the address generation is:

```
for (m1 = 0 to M’/Mu-1)
    for (n1 = 0 to N’/Nu-1)
        for (k1 = 0 to K’/Ku-1)
            parfor(0 to Mu)
            parfor(0 to Ku)
              addr_o[Mu-1:0][Ku-1:0] = ptr_o + temporal_address + spatial_address
              temporal_address = m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1
              spatial_address = [Mu-1:0]*s_str_m + [Ku-1:0]*str_str_k

```
where t_str_* - are temporal strides and s_str_* - are spatial strides.

> Which actually translates to:

Reviewer 3: ```suggestion
Which translates to:
```

```
addr_o[0][0] = ptr_o + m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1 + [0]*s_str_m + [0]*str_str_k
addr_o[0][1] = ptr_o + m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1 + [0]*s_str_m + [1]*str_str_k
addr_o[0][2] = ptr_o + m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1 + [0]*s_str_m + [2]*str_str_k
…
addr_o[m][k] = ptr_o + m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1 + [m]*s_str_m + [k]*str_str_k
…
addr_o[Mu-1][Ku-3] = ptr_o + m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1 + [Mu-1]*s_str_m + [Ku-3]*str_str_k
addr_o[Mu-1][Ku-2] = ptr_o + m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1 + [Mu-1]*s_str_m + [Ku-2]*str_str_k
addr_o[Mu-1][Ku-1] = ptr_o + m1*t_str_m1 + n1*t_str_n1 + k1*t_str_k1 + [Mu-1]*s_str_m + [Ku-1]*str_str_k
```

## IO ports
The input and output ports of the Streamer are shown in the table below.

> The Streamer uses a simplified CSR request/response interface for CSR write/read operation. A more detailed description of the CSR operation interface can be found at [here](https://kuleuven-micas.github.io/snitch_cluster/rm/snax_cluster.html).

Reviewer: ```suggestion
The Streamer uses a simplified CSR request/response interface for CSR write/read operation. A more detailed description of the CSR operation interface can be found [here](https://kuleuven-micas.github.io/snitch_cluster/rm/snax_cluster.html).
```

> The Streamer uses a simplified TCDM request/response interface to read and write data from/to the TCDM. A more detailed description of the TCDM request/response interface can be found at [here](https://kuleuven-micas.github.io/snitch_cluster/rm/snax_cluster.html).

Reviewer: ```suggestion
The Streamer uses a simplified TCDM request/response interface to read and write data from/to the TCDM. A more detailed description of the TCDM request/response interface can be found [here](https://kuleuven-micas.github.io/snitch_cluster/rm/snax_cluster.html).
```

> The Streamer uses the Decoupled interface for accelerator input and output data. A more detailed description of the Decoupled interface can be found at [here](https://www.chisel-lang.org/docs/explanations/interfaces-and-connections#the-standard-ready-valid-interface-readyvalidio--decoupled).

Reviewer: ```suggestion
The Streamer uses the Decoupled interface for accelerator input and output data. A more detailed description of the Decoupled interface can be found [here](https://www.chisel-lang.org/docs/explanations/interfaces-and-connections#the-standard-ready-valid-interface-readyvalidio--decoupled).
```

| | . | . | . | . | The accelerator2streamer ports number is the same as data writer number. The index 0 corresponds to the first output data and 2 for second output data (if any) and so on.|
| | | | | | | |

### CSR definition
| Address | CSR name             | Notes                               |
|---------|--------------------------|-------------------------------------|
| offset* + [0..temporalDim - 1]   |    temporalLoopBoundCSRs     | temporal loop bound for each temporal dimension.           |
| offset + temporalDim + [0..temdataMoverNum * temporalDimporalDim - 1]   |    temporalLoopSrtidesCSRs     | temporal loop strides for each temporal dimension and for each data mover.          |
| offset + temporalDim +  dataMoverNum * temporalDim + [0..spatialDim.sum - 1]   |    spatialLoopSrtidesCSRs     | spatial loop strides for each data mover and for corresponding spatial dimension. The spatial dimension for each data mover can be different. It depends on the accelerator.        |
| offset + temporalDim +  dataMoverNum * temporalDim + spatialDim.sum + [0..dataMoverNum - 1]   |    basePtrCSRs     |  base pointers for each data mover.          |

> *offset is defined by the SNAX core. A more detailed explanation of what are these configurations can be found at `StreamerTop.scala`.

Reviewer 2: maybe don't use the * here to reference the offset, this is a bit confusing, since it might be interpreted as a pointer dereference operator and it is used as  a multiplication sign further down the line

Reviewer 2: Maybe just put the explanation of the offset before the table

Author: Ok!
